lib/oxigraph/src/sparql/dataset.rs

```diff
@@ -1,18 +1,235 @@
+    let hdt_str = match term {
+        // Remove double quote delimiters from URIs.
+        Term::NamedNode(named_node) => named_node.clone().into_string(),
+
+        // Get the string directly from literals and add
+        // quotes to work-around handling of "\n" being
+        // double-escaped.
+        // format!("\"{}\"", literal.value()),
+        Term::Literal(literal) => {
+            if literal.is_plain() {
+                literal.to_string().replace("\\n", "\n")
+            }
+            // For numbers and other typed literals return
+            // None as the BGP search will need to collect
+            // all possibilities before filtering.
+            else {
+                return Err(StorageError::Corruption(CorruptionError::new(format!(
+                    "unhandled literal value {literal}"
+                ))));
+            }
+        }
+
+        Term::BlankNode(_s) => term.to_string(),
+
+        // Otherwise use the string directly.
+        Term::Triple(_) => term.to_string(),
+    };
+
+    Ok(hdt_str)
+}
+
+impl QueryableDataset for HDTDatasetView {
+    type InternalTerm = String;
+    type Error = StorageError;
+
+    fn internal_quads_for_pattern(
+        &self,
+        subject: Option<&String>,
+        predicate: Option<&String>,
+        object: Option<&String>,
+        graph_name: Option<Option<&String>>,
+    ) -> Box<dyn Iterator<Item = Result<InternalQuad<Self>, StorageError>>> {
+        if let Some(graph_name) = graph_name {
+            if graph_name.is_some() {
+                return Box::new(
+                    vec![Err(StorageError::Corruption(CorruptionError::new(
+                        format!("HDT does not support named graph: {graph_name:?}"),
+                    )))]
+                    .into_iter(),
+                );
+            }
+        }
+        let s = subject.cloned();
+        let p = predicate.cloned();
+        let o = object.cloned();
+
+        // Create a vector to hold the results.
+        let mut v: Vec<Result<InternalQuad<_>, StorageError>> = Vec::new();
+
+        for data in &self.hdts {
+            // Query HDT for BGP by string values.
+            let results = data
+                .hdt
+                .triples_with_pattern(s.as_deref(), p.as_deref(), o.as_deref());
+
+            // For each result
+            for result in results {
+                let ex_s = (*result.0).to_string();
+                let ex_p = (*result.1).to_string();
+                let ex_o = (*result.2).to_string();
+
+                // Add the result to the vector.
+                v.push(Ok(InternalQuad {
+                    subject: ex_s,
+                    predicate: ex_p,
+                    object: ex_o,
+                    graph_name: None,
+                }));
+            }
+        }
+
+        Box::new(v.into_iter())
+    }
+
+    fn internalize_term(&self, term: Term) -> Result<String, StorageError> {
+        term_to_hdt_bgp_str(&term)
```

Comment:
I think you can convert `term_to_hdt_bgp_str` to take an owned `Term` and not a reference and avoid the `.clone()` call for NamedNodes.